packages/neuron-wallet/src/services/addresses.ts

```diff
@@ -149,128 +160,196 @@ export default class AddressService {
+        codeHash: SystemScriptInfo.SECP_CODE_HASH,
+        hashType: SystemScriptInfo.SECP_HASH_TYPE,
+        args: address.blake160
+      });
+      const lockHash = script.computeHash()
+      const txCount = txCountsByLockHashes.get(lockHash)
+
+      if (!txCount) {
+        continue
+      }
+      address.txCount = txCount
     }
-    return addressEntity
+
+    const allUnusedAddresses = addresses.filter(address => !address.txCount)
+
+    return allUnusedAddresses
   }
 
-  public static allAddresses = (): AddressInterface[] => {
-    return AddressDao.allAddresses( AddressService.getAddressVersion())
+  public static async nextUnusedAddress (walletId: string): Promise<AddressInterface | undefined> {
+    const [unusedReceivingAddresses, unusedChangeAddresses] = await this.getUnusedAddresses(walletId)
+    if (unusedReceivingAddresses.length) {
+      return unusedReceivingAddresses[0]
+    }
+    if (unusedChangeAddresses.length) {
+      return unusedChangeAddresses[0]
+    }
+
+    return undefined
   }
 
-  public static allAddressesByWalletId = (
-    walletId: string,
-    addressVersion: AddressVersion = AddressService.getAddressVersion()
-  ): AddressInterface[] => {
-    return AddressDao.allAddressesByWalletId(walletId, addressVersion)
+  public static async nextUnusedChangeAddress (walletId: string): Promise<AddressInterface | undefined> {
+    const [ , unusedChangeAddresses ] = await this.getUnusedAddresses(walletId)
+    if (unusedChangeAddresses.length) {
+      return unusedChangeAddresses[0]
+    }
+    return undefined
   }
 
-  public static allBlake160sByWalletId(walletId: string): string[] {
-    return AddressService.allAddressesByWalletId(walletId).map(addr => addr.blake160)
+  public static async allUnusedReceivingAddresses(walletId: string): Promise<AddressInterface[]> {
+    const [ unusedReceivingAddresses ] = await this.getUnusedAddresses(walletId)
+    return unusedReceivingAddresses
   }
 
-  public static allLockHashes(): string[] {
-    const addresses = AddressService.allAddresses().map(address => address.address)
-    return AddressParser.batchToLockHash(addresses)
+
+  public static allAddresses = async (): Promise<AddressInterface[]> => {
+    const publicKeyInfos = await AddressDao.allAddresses()
+    return publicKeyInfos.map(publicKeyInfo => ({
+      walletId: publicKeyInfo.walletId,
+      address: publicKeyInfo.address,
+      path: publicKeyInfo.path,
+      addressIndex: publicKeyInfo.addressIndex,
+      addressType: publicKeyInfo.addressType,
+      blake160: publicKeyInfo.publicKeyInBlake160,
+      description: publicKeyInfo.description,
+    }))
   }
 
-  public static allLockHashesByWalletId(walletId: string): string[] {
-    const addresses = AddressService.allAddressesByWalletId(walletId).map(addr => addr.address)
-    return AddressParser.batchToLockHash(addresses)
+  public static async allAddressesByWalletId (walletId: string): Promise<AddressInterface[]> {
```

Comment:
Hard to tell the difference between `allAddresses` and `allAddressesByWalletId` by their names.